clusterloader/docs/vision.md

```diff
@@ -0,0 +1,320 @@
+# Cluster loader vision
+
+Author: wojtek-t
+
+Last update time: 9th May 2018
+
+## Background
+
+As of 31/03/2018, all our scalability tests, are regular e2e tests written in Go.
+It makes them really unfriendly for developers not working on scalability who just
+want to load tests Kubernetes features they are working on. Doing so in many
+situation requires understanding how those tests really work, modify their code
+to test the new feature and only then run and debug tests (or create load test for
+your feature on your own, which may be easier, but on the other hand may not
+exercise important metrics that our performance test check). This workflow is far
+from optimal.
+
+That said, long time ago we came up with the idea that users should be able to
+just bring their own objects definitions in JSON format, potentially annotate them
+with some metadata to describe how the load should be generated from these and
+testing infrastructure should do everything else automatically.
+
+In early 2017 the prototype of “Cluster Loader” has been created. It proved that
+configuring tests with json/yaml files is possible. But its functionality is very
+limited and it was very far from enabling migration any of existing scalability
+tests to that framework.
+
+We would like to get back to that and build fully functional Cluster Loader and
+use it as a framework to run all our scalability tests. This doc is describing
+high-level vision of how this will work.
+
+
+## Vision
+
+At the high level, a single test will consist of a number of phases (if we have
+really good usecase we may consider allowing the to create a DAG, but for now
+I don’t have rationale for it). In each phase we will be creating/updating/deleting
+a number of different objects.
+Additionally, we will introduce a set of predefined operations (that user will be
+able to use as phases). They will allow users to monitor/measure performance impact
+of user-defined phases.
+The following subsections will describe this in a bit more details.
+
+
+### Tuning Set
+
+Since we would like to be fully load even very big clusters, we need to be able to
+create a number of “similar” objects. A “Tuning Set” concept will allow us to
+spread those operations across some time.
+We define Tuning Set as following:
+
+```
+struct TuningSet {
+	Name string
+	// Exactly one of the following should be set.
+	QpsLoad        *QpsLoad
+	RandomizedLoad *RandomizedLoad
+	SteppedLoad    *SteppedLoad
+}
+
+// QpsLoad defines a uniform load with a given QPS.
+struct QpsLoad {
+	InitialDelay time.Duration
+	Qps          float
+}
+
+// RandomizedLoad defines a load that is spread randomly
+// across a given total time.
+struct RandomizedLoad {
```

Comment:
Can you help me understand how this is different from QpsLoad?

Reply:
QPSLoad is that we generate exactly "QPS" queries per second (so one query every 1/QPS seconds).
RandomizedLoad is that we randomize the load over such period of time.

They are pretty similar in general.

[As an example:
- RandomizedLoad is how we spread over time things in load test
- QpsLoad is how we spread over time creations of latency pods]

Reply:
I see.. thanks.

In that case, I believe it'll be much better to have `TimeSpread` instead of `AverageQps` (as you probably don't want to calculate it's value based on TimeSpread and #calls).

Reply:
I'm not sure I understand it.
I think that average qps is exactly what I would like to pass.

IIUC, `TimeSpread` you're proposing correctly, would mean that I would like to create everything over a given time - this for me would be a separate policy (which we may want to add in the future, but for the purpose of migrating our existing e2e tests it won't be needed)

Reply:
Ok.. let me instead ask this question. What do you think should be the averageQps for the scaling operation in our load test?

Reply:
Currently we have exactly that - it's by default set to 10.

I don't want to use timespread, because then if you say change the size of the cluster - they you only need to change the number of objects and not change the tunning set.

Reply:
> That's the expected throughput we want for pod churn. I believe what we want here is the rate at which we're going to change the RC objects as those are what would be involved in the Phase operations IIUC.

Yes - that was kind of shortcut from my side.
So let's assume that we want to have 10pods/s churn on average.
Then, if small RCs are 1/2 of all pods, then for these we want to generate 5pods/s. Given they have numReplicas=5, we would like to generate 1 per s on average.
if medium RCs are 1/4 of all pods, then we want to generate 2.5pods/s. Given they have numReplicas=25 (IIRC), that would mean an average load of 0.1 per s.
and so on

Yes - it requires some match to be done by someone creating a test. We may consider changing our tests in the future. But I think the steps should be:
- migrate to cluster loader
- then start changing tests if we want
We shouldn't do the two steps at the same time.

Reply:
Makes sense.. thx for explaining.